### ISSUE TITLE
Option to clear the REPL like DrRacket does

### ISSUE THREAD

**Issue by greghendershott** [greghendershott/racket-mode#149]
As suggested by acarrico on `#racket`.  On each `racket-run`, clear the `*Racket REPL*` buffer of old output.  (In that case, perhaps do `(banner)` to show Racket version info each time, like DrRacket does, too?)

I actually like the status quo behavior (surprise :)). I like to accumulate history; sometimes I'll refer back to it. If/when it becomes too much, I just <kbd>C-x h</kbd>and <kbd>C-w</kbd>.

However I can understand someone preferring the proposed behavior. Probably it would be easy to add as an option.

---

Meanwhile, people could add this behavior in their Emacs init file by advising `racket-run`. Although I never use advice so I'm not sure the following is correct, it seemed to work:

``` cl
(require 'racket-mode)
(defadvice racket-run (before my/before-racket-run)
  "Advice to clear the Racket REPL buffer before each `racket-run', like DrRacket."
  (pcase (get-buffer "*Racket REPL*")
    (`() nil)
    (buf (with-current-buffer buf
           (delete-region (point-min) (point-max))))))
(ad-activate #'racket-run)
```

**Comment by acarrico** [comment]
Yes, it would be something I'd like to shut on and off depending on the current workflow.

**Comment by acarrico** [comment]
I think this feature request is less necessary now that I've tried to get in the habit of using C-x ` to go through errors.

**Comment by jeremyheiler** [comment]
+1

I'd also like to have it exposed as a command. I often clear my Clojure REPL when I'm done with whatever task I am doing. Though, it might be less necessary if it's cleared per run.

**Comment by greghendershott** [comment]
I'm actually kind of leaning towards doing neither of these.

- **Clearing before every Run:** It sounds like @acarrico doesn't really want this anymore. For people who do want it, adding the `defadvice` I showed above to `.emacs` does work.
 
- **Clearing it manually, when desired:** It's already possible to clear the `*Racket REPL*` buffer in the usual Emacs way:

    - <kbd>C-x h</kbd> does `mark-whole-buffer` -- what younger apps call "Select All" :)
    - <kbd>C-w</kbd> does `kill-region` -- "Cut"

    A new `racket-clear-repl` command would be... only that. It would be super easy for me to add, but I feel like I'd be adding redundancy not value?

**Comment by acarrico** [comment]
You are right that duping racket-clear-repl would be redundant, however I presume the new command would work from the source code's buffer too, however I think you are correct that it isn't worth cluttering up the keybindings.  I think I will experiment with adding advice to racket-run to see how I like it.

**Comment by acarrico** [comment]
(... moments later) I added the advice and I DO like it. I hit F5 and seeing the buffer immediately clear feels good. I think this is especially nice becoause typed racket is SO SLOW and there is no feedback. Now, when I see an error, or the prompt, I know it is done. Sometimes I'm not sure without the advice, especially if the old error comes up and it scrolls to the same place, or a new prompt comes under the old ones. Is it done? Did it do anything?

**Comment by greghendershott** [comment]
> I added the advice and I DO like it.

Cool.

> You are right that duping racket-clear-repl would be redundant, however I presume the new command would work from the source code's buffer too,

Oh, good point. <kbd>C-c C-z C-x h C-w C-c Cz</kbd> is a lot. I could say "Emacs macros!", but, if it's a common task for enough people, it could be worth adding.

That's been an interesting aspect of doing racket-mode, understanding and appreciating the different workflows that people have. That reminds me of doing Cakewalk software for musicians. And worries me. Pleasing all the people all the time is, in my experience, undecidable. :P

It's also interesting wondering when its more helpful to explain how to both "DIY" and "DRY" with general Emacs techniques, vs. provide bespoke commands. To be clear, I don't say that like I'm any kind of Emacs expert giving people fishing nets. On the contrary, I've put things in racket-mode only to discover later that, yeah, that's already built into Emacs for 20 years now, rip it out, and wire things up to that standard thing. So yeah. :)

**Comment by greghendershott** [comment]
@acarrico On `#racket` I saw you mention that `next-error` isn't working with the `defadvice` above. I can repro this. In fact, same issue just manually clearing it with `kill-region`.

I've spent a couple hours trying to understand how Emacs' `compile.el` works in this regard. I think I've tracked it down to how `compilation-forget-errors` sets `compilation-messages-start`: It is set to a `marker`, which means the marker "floats down", staying below newly-inserted text. Therefore the error messages are ignored, because they're above the "start" marker.

This behavior doesn't seem quite right. From the source comments, it looks like it may have changed, and maybe recently (? I"m on 25.0.50.2 -- how about you?). But I'll need to dig into Emacs git history to figure out when/why. Maybe it's a bug that needs to be fixed in Emacs?  But even if so, we need a work-around meanwhile.

I tried setting `compilation-messages-start` to `nil` when the buffer is empty. That fixes the problem. But I'm not sure if it's correct. Even if it's correct, I'm not sure if it's sufficient to solve similar problems we haven't noticed yet.

TL;DR I have a work-around. I'll do more research to determine if I ought to commit and push that, or, do something else.

**Comment by acarrico** [comment]
GNU Emacs 24.5.1 (x86_64-unknown-linux-gnu, GTK+ Version 3.16.6)

**Comment by acarrico** [comment]
Also, just to add to the confusion, it doesn't clear the REPL anymore. At least not in the buffers I'm currently editing.  I haven't restarted emacs. To add the advice, I originally hit C-j after each of the following lines in *scratch*. Have I lost my advice? Or is it only applying to certain buffers--this is global isn't it?

(defadvice racket-run (before my/before-racket-run)
  "Advice to clear the Racket REPL buffer before each `racket-run', like DrRacket."
  (pcase (get-buffer "*Racket REPL*")
    (buf (with-current-buffer buf
           (delete-region (point-min) (point-max))))))

(ad-activate #'racket-run)

**Comment by acarrico** [comment]
Oh, actually forget my last comment. I only have advice on racket-run. The behavior I was reporting was just me expecting C-c C-t to clear the REPL, but of course I don't have advice on racket-test.

**Comment by greghendershott** [comment]
> I tried setting `compilation-messages-start` to nil when the buffer is empty. That fixes the problem. But I'm not sure if it's correct. Even if it's correct, I'm not sure if it's sufficient to solve similar problems we haven't noticed yet.

I did find a similar problem: When you start the Racket REPL for the very first time on a file that contains an error, `next-error` doesn't work. Same ultimate reason, `compilation-messages-start` ends up "below" the error messages. But can't use a simple "buffer is empty" test, in this case.

I have a another work-around but was hitting an edge case that puzzled me, before I had to give up last night. Will try again, soon.

**Comment by greghendershott** [comment]
@acarrico I split out the `next-error` problem as issue #166. I just pushed a commit to fix.

**Comment by acarrico** [comment]
I got a new version from melpa to test out issue #166 and it seems good if I add the advice with C-j in the \*scratch\* buffer. However, when giving the advice in .emacs/init.el, and hitting <F5>, emacs  complains, "Wrong type argument: stringp, nil". Does this mean I need to load racket mode somehow before adding the advice in my init.el?

**Comment by acarrico** [comment]
It seems like I actually need to do one racket-run before adding the advice.

**Comment by greghendershott** [comment]
What if you `(require 'racket-mode)` in your init.el prior to the `defadvice` referencing `racket-run`?

**Comment by acarrico** [comment]
Yes, so now we have a nice recipe :).

**Comment by greghendershott** [comment]
Great! I just edited my original issue description at the top of this thread, to add the `(require 'racket-mode)`.

[Although that makes these last few comments look odd, I figure on balance it's less confusing in case someone just reads the first item and tries to use that code as-is.]

**Comment by acarrico** [comment]
As usual, I'm finding every pitfall. After adding
```
(require 'racket-mode)
```
the advice wasn't actually applied, my init was actually bailing. I did figure out the answer though, I was missing
```
 (package-initialize)
```
in this bit of initialization magic:
```
(require 'package)
(add-to-list 'package-archives
  '("melpa" . "http://melpa.milkbox.net/packages/") t)
(package-initialize)
```

**Comment by greghendershott** [comment]
Oh, yeah, AFIK just generally you ought to call `package-initialize`.

---

BTW I **love** [`use-package`](https://github.com/jwiegley/use-package) to organize my init.el. So I have to manually install _that_ `use-package` package :) -- but no others. So I have:

```
(require 'package)
(setq package-enable-at-startup nil) ;; for `use-package' :defer
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(package-initialize)
(unless (package-installed-p 'use-package)
  (message "use-package not installed. Trying to install")
  (package-refresh-contents)
  (package-install 'use-package))
(require 'use-package)
(require 'bind-key)
```

Then the bulk of my init.el is mainly forms like this, where the `:ensure t` tells `use-package` to install the package if necessary.

```
(use-package yaml-mode
  :ensure t
  :mode "\\.yml\\'")
```

**Comment by acarrico** [comment]
One more issue solved. Note the (and buf ...) in this version:

(require 'racket-mode)
(defadvice racket-run (before my/before-racket-run)
  "Advice to clear the Racket REPL buffer before each `racket-run', like DrRacket."
  (pcase (get-buffer "*Racket REPL*")
    (buf (and buf
              (with-current-buffer buf
                (delete-region (point-min) (point-max)))))))
(ad-activate #'racket-run)

**Comment by greghendershott** [comment]
Ah. Sorry! That `pcase` is wrong in the same way that a Racket `match` like that would be wrong.It's missing a check for `nil`. I'd meant something like:

```cl
 (pcase (get-buffer "*Racket REPL*")
    ('nil nil)  ;; <=======
    (buf (with-current-buffer buf
           (delete-region (point-min) (point-max)))))
```

Or the way you wrote it is fine. Or with a `let`. Or whatever. I'm sorry my code example was wrong, and glad you figured it out!

**Comment by acarrico** [comment]
Zombie issue 149 is alive!!!!!! As far as I can tell, that 'nil UPATTERN must be `nil.

**Comment by acarrico** [comment]
I'm almost afraid of what will happen next time I restart emacs. Still loving the elisp?

**Comment by acarrico** [comment]
BTW, maybe put this in the examples dir?

**Comment by greghendershott** [comment]
The pattern might need to be `` `() `` to work on both Emacs 24.3 and 25. At least that's what I recently discovered and did with commit d779a24.

I'm spoiled by Racket `match` and I'm also not loving `nil` punning.

**Comment by greghendershott** [comment]
I updated the example in the first comment here, to use `` `() ``.

I should put this in the README ... assuming I've finally gotten it right. O_o

(BTW The `examples` dir is probably not the best name. It originated as example Racket code to see the font-lock and indent. But now the files there are actually used as part of tests. Anyway it's examples of Racket not examples of Elisp to configure racket-mode.)

**Comment by sorawee** [comment]
I recently need this functionality, but `delete-region`  no longer works. Instead, I need:

```
(defun my/racket-repl-clear ()
  (interactive)
  (let ((inhibit-read-only t))
    (erase-buffer)))
```

In case it's helpful to anyone who comes across.

**Comment by greghendershott** [comment]
`comint-kill-region` -- which `racket-repl-mode` binds to <kbd>C-w</kbd> by default -- should kill the read-only prompts, too.

Interactively, you could use that after `mark-whole-buffer` (<kbd>C-x h</kbd>).

My muscle memory has <kbd>C-x h C-w</kbd>.

---

From Elisp, you can just directly supply the region beginning/end values as arguments:

```elisp
(defun my/racket-repl-clear ()
  (interactive)
  (comint-kill-region (point-min) (point-max)))
```

Not that what you have is wrong.

**Comment by Bost** [comment]
@greghendershott it looks like the `comint-kill-region` also deletes the REPL prompt "> ", although it shouldn't. Hmm :( So then a hackey `delete-region` with `(- (point-max) prompt-length)` would do the job:
```elisp
  (defun my/racket-repl-clear ()
    (interactive)
    (let ((inhibit-read-only t))
      (delete-region (point-min) (- (point-max) 2))))
```
However, I think even this hack can't be done properly. It looks like the prompt string is hard coded in the [racket/interactions.rkt](https://github.com/greghendershott/racket-mode/blob/abd59fd4538f09fb201957d87887e58e377f0255/racket/interactions.rkt#L57):
```racket
(define (display-prompt str)
  (flush-output (current-error-port))
  (fresh-line)
  (display str)
  (display "> ")
  (flush-output)
  (zero-column!))
```
And that's odd. Or am I misinterpreting something? (My racket-fu is rather young). 

OTOH in our particular case, the `delete-region` may be more appropriate than the `comint-kill-region` since it doesn't modify the kill ring.

**Comment by greghendershott** [comment]
@Bost In the context of the original issue, it's fine that the last prompt is erased, too, because the run command will display another prompt afterwards.

**Comment by greghendershott** [comment]
But I can also imagine some people wanting a command that deletes the REPL buffer _except_ for the last prompt (if any).

I have a commit for all of this I might merge.

**Comment by greghendershott** [comment]
I merged a commit for this.

There are two new hooks, [racket-before-run-hook](https://www.racket-mode.com/#racket_002dbefore_002drun_002dhook) and [racket-after-run-hook](https://www.racket-mode.com/#racket_002dafter_002drun_002dhook).

There is also a function [racket-repl-clear](https://www.racket-mode.com/#racket_002drepl_002dclear), which you can use with `racket-before-run-hook`.

So either use `M-x customize` to set this, or in your Emacs init file do something like `(add-hook ‘racket-before-run-hook #’racket-repl-clear)`.

---

The commit is a bit more complicated than you might expect (and more than I expected going into this). For one thing, I got annoyed by there being a blank line at the start of the before, when it was cleared. That led me to review some code for a function called `fresh-line` prints a newline _unless_ already in column zero. And yada yada yada.

Anyway I think the result is a bit better than people could have done already on their own, so I hope it's helpful.
